File: mongo/go/findone/findone.go
// MongoDB db.collection.findOne() method examples in Golang

package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

const uri = "mongodb://bigboxuser:bigboxpass@localhost:27017/?maxPoolSize=20&w=majority"

func main() {
	// Initiate mongo client
	mongoClient, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(uri))
	if err != nil {
		fmt.Println("MongoDB connection error: ", err)
		os.Exit(1)
	}
	defer mongoClient.Disconnect(context.Background())

	// Initiate collection instance
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	customerCollection := mongoClient.Database("bigboxcode").Collection("customer")
	wrongCollection := mongoClient.Database("bigboxcode").Collection("wrong_collection")

	// Insert some documents
	insertResult, err := customerCollection.InsertMany(ctx, []interface{}{
		map[string]interface{}{
			"name":  "Izabella Conroy",
			"age":   19,
			"phone": "108-679-2154",
			"address": map[string]interface{}{
				"street":     "11 North Vineyard Drive",
				"city":       "Pueblo West",
				"state":      "CO",
				"postalCode": "81007",
			},
		},
		map[string]interface{}{
			"name":  "Alisa Parker",
			"age":   72,
			"phone": "768-319-1054",
			"address": map[string]interface{}{
				"street":     "8532 Ingalls Circle",
				"city":       "Arvada",
				"state":      "CO",
				"postalCode": "80003",
			},
		},
		map[string]interface{}{
			"name":  "Eloise Weber",
			"age":   29,
			"phone": "618-357-2104",
			"address": map[string]interface{}{
				"street":     "632 Belmar Drive",
				"city":       "Edmond",
				"state":      "OK",
				"postalCode": "73025",
			},
		},
	})

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.insertMany() | Result: ", insertResult)

	var commandResult bson.M

	// Find one document from cutomer collection
	err = customerCollection.FindOne(ctx, bson.M{}).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne() | Result: ", commandResult)

	// Apply fitler where age=29
	err = customerCollection.FindOne(ctx, bson.M{"age": 29}).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({age: 29}) | Result: ", commandResult)

	// Apply filter where age >= 29
	err = customerCollection.FindOne(ctx, bson.M{"age": bson.M{"$gte": 29}}).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({age: {$gte: 29}}) | Result: ", commandResult)

	// Use $or operator for the query filter
	filter := bson.M{"$or": []bson.M{{"age": bson.M{"$gt": 29}}, {"address.postalCode": "81007"}}}
	err = customerCollection.FindOne(ctx, filter).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({$or: [{age: {$gt: 29}}, {\"address.postalCode\": \"81007\"}]}) | Result: ", commandResult)

	// Set projection to get name, phone, and postalCode only
	findOptions := options.FindOne().SetProjection(bson.M{"name": 1, "phone": true, "address.postalCode": 1})
	err = customerCollection.FindOne(ctx, bson.M{"age": 29}, findOptions).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({age: 29}, { name: 1, phone: true, \"address.postalCode\": 1}) | Result: ", commandResult)

	// Ignore properties
	findOptions = options.FindOne().SetProjection(bson.M{"name": false, "phone": 0, "address.city": 0})
	err = customerCollection.FindOne(ctx, bson.M{}, findOptions).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({}, { name: false, phone: 0, \"address.city\": 0}) | Result: ", commandResult)

	// Apply filter that doesn't return any result
	err = customerCollection.FindOne(ctx, bson.M{"age": 299999999}).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({age: 299999999}) | Result: ", commandResult)

	// Use regex to filter by matching value of a field
	regexPattern := primitive.Regex{Pattern: "^alisa", Options: "i"}
	err = customerCollection.FindOne(ctx, bson.M{"name": bson.M{"$regex": regexPattern}}).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({name: {$regex : /^alisa/i}}) | Result: ", commandResult)

	// Use regex
	regexPattern = primitive.Regex{Pattern: "drive", Options: "i"}
	err = customerCollection.FindOne(ctx, bson.M{"address.street": bson.M{"$regex": regexPattern}}).Decode(&commandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.customer.findOne({\"address.street\": {$regex : /drive/i}}) | Result: ", commandResult)

	// Try to use findOne() one a non existing collection, We get null
	var wrongCommandResult bson.M
	err = wrongCollection.FindOne(ctx, nil).Decode(&wrongCommandResult)

	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Command: db.wrong_collection.findOne() | Result: ", wrongCommandResult)

}

File: mongo/JavaMongo/src/FindOne.java
// db.collection.findOne() method example in Java

import com.mongodb.client.*;
import com.mongodb.client.model.Filters;
import org.bson.Document;

import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

public class FindOne {

    public static void main(String[] args) {
        String uri = "mongodb://bigboxuser:bigboxpass@localhost:27017/";

        try (MongoClient mongoClient = MongoClients.create(uri)) {
            MongoDatabase database = mongoClient.getDatabase("bigboxcode");
            MongoCollection<Document> customerCollection = database.getCollection("customer");
            MongoCollection<Document> wrongCollection = database.getCollection("wrong_collection");

            // Insert some documents
            List<Document> documents = Arrays.asList(
                    new Document("name", "Izabella Conroy")
                            .append("age", 19)
                            .append("phone", "108-679-2154")
                            .append("address", new Document("street", "11 North Vineyard Drive")
                                    .append("city", "Pueblo West")
                                    .append("state", "CO")
                                    .append("postalCode", "81007")),
                    new Document("name", "Alisa Parker")
                            .append("age", 72)
                            .append("phone", "768-319-1054")
                            .append("address", new Document("street", "8532 Ingalls Circle")
                                    .append("city", "Arvada")
                                    .append("state", "CO")
                                    .append("postalCode", "80003")),
                    new Document("name", "Eloise Weber")
                            .append("age", 29)
                            .append("phone", "618-357-2104")
                            .append("address", new Document("street", "632 Belmar Drive")
                                    .append("city", "Edmond")
                                    .append("state", "OK")
                                    .append("postalCode", "73025"))
            );

            customerCollection.insertMany(documents);
            System.out.println("Command: db.customer.insertMany() | Result: " + documents);


            // Find one document from cutomer collection
            Document findOneResult = customerCollection.find().first();

            System.out.println("Command: db.customer.findOne() | Result: " + findOneResult.toString());

            // Apply fitler where age=29
            findOneResult = customerCollection.find(Filters.eq("age", 29)).first();

            System.out.println(
                    "Command: db.customer.findOne({age: 29}) | Result: " + findOneResult.toString());

            // Apply filter where age >= 29
            findOneResult = customerCollection.find(Filters.gte("age", 29)).first();

            System.out.println(
                    "Command: db.customer.findOne({age: {$gte: 29}}) | Result: " + findOneResult.toString());

            // Use $or operator for the query filter
            findOneResult = customerCollection.find(Filters.or(Filters.gt("age", 29), Filters.eq("address.postalCode", "81007"))).first();

            System.out.println(
                    "Command: db.customer.findOne({$or: [{age: {$gt: 29}}, {\"address.postalCode\": \"81007\"}]}) | Result: " + findOneResult.toString());

            // Set projection to get name, phone, and postalCode only
            findOneResult = customerCollection.find(Filters.eq("age", 29)).projection(new Document("name", 1).append("phone", true).append("address.postalCode", 1)).first();

            System.out.println("Command: db.customer.findOne({age: 29}, { name: 1, phone: true, \"address.postalCode\": 1}) | Result: " + findOneResult.toString());

            // Ignore properties
            findOneResult = customerCollection.find().projection(new Document("name", false).append("phone", 0).append("address.city", 0)).first();

            System.out.println("Command: db.customer.findOne({}, { name: false, phone: 0, \"address.city\": 0}) | Result:" + findOneResult.toString());

            // Apply filter that doesn't return any result
            findOneResult = customerCollection.find(Filters.eq("age", 299999999)).first();

            System.out.println("Command: db.customer.findOne({age: 299999999}) | Result: " + findOneResult);

            // Use regex to filter by matching value of a field
            Pattern pattern = Pattern.compile("^alisa", Pattern.CASE_INSENSITIVE);
            findOneResult = customerCollection.find(Filters.regex("name", pattern)).first();

            System.out.println(
                    "Command: db.customer.findOne({name: {$regex : /^alisa/i}}) | Result: " + findOneResult.toString());

            // Use regex
            pattern = Pattern.compile("drive", Pattern.CASE_INSENSITIVE);
            findOneResult = customerCollection.find(Filters.regex("address.street", pattern)).first();

            System.out.println("Command: db.customer.findOne({\"address.street\": {$regex : /drive/i}}) | Result: " + findOneResult.toString());

            // Try to use findOne() one a non existing collection, We get null
            findOneResult = wrongCollection.find().first();

            System.out.println("Command: db.wrong_collection.findOne() | Result: " + findOneResult);

        }
    }
}
